drv/sidecar-mainboard-controller/src/fan_modules.rs

```diff
@@ -0,0 +1,154 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::{Addr, MainboardController, Reg};
+use bitfield::bitfield;
+use drv_fpga_api::{FpgaError, FpgaUserDesign, WriteOp};
+use userlib::FromPrimitive;
+use zerocopy::{AsBytes, FromBytes};
+
+bitfield! {
+    #[derive(Copy, Clone, PartialEq, Eq, FromPrimitive, AsBytes, FromBytes)]
+    #[repr(C)]
+    pub struct FanModuleStatus(u8);
+    pub enable, set_enable: 0;
+    pub led, set_led: 1;
+    pub present, _: 2;
+    pub power_good, _: 3;
+    pub power_fault, _: 4;
+    pub power_timed_out, _: 5;
+}
+
+/// Each fan module contains two individually controlled fans
+pub const NUM_FAN_MODULES: usize = 4;
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum FanModuleLedState {
+    Off = 0,
+    On = 1,
+    Blink = 2,
+}
+
+pub struct FanModules {
+    fpga: FpgaUserDesign,
+    presence: [bool; NUM_FAN_MODULES],
+    led_state: [FanModuleLedState; NUM_FAN_MODULES],
+}
+
+impl FanModules {
+    pub fn new(task_id: userlib::TaskId) -> Self {
+        Self {
+            fpga: FpgaUserDesign::new(
+                task_id,
+                MainboardController::DEVICE_INDEX,
+            ),
+            led_state: [FanModuleLedState::Off; NUM_FAN_MODULES],
+            presence: [false; NUM_FAN_MODULES],
+        }
+    }
+
+    /// Fetch the FANx_STATE register from the FPGA for all fan modules
+    ///
+    /// Additionally, update internal presence state with the latest status.
+    pub fn get_status(
+        &mut self,
+    ) -> Result<[FanModuleStatus; NUM_FAN_MODULES], FpgaError> {
+        let status: [FanModuleStatus; NUM_FAN_MODULES] =
+            self.fpga.read(Addr::FAN0_STATE)?;
+        for (module, status) in status.iter().enumerate() {
+            self.presence[module] = status.present();
+        }
```

Comment:
Nit: you could avoid this loop with
```suggestion
        self.presence = status.map(|s| s.present());
```